symintify all of derivatives.yaml

bdhirsh opened the pull request:
Big-bang PR to symintify **all** .sizes() calls in derivatives.yaml, which will be needed for symbolic tracing.

* with the exception of `split()`, which is tougher to land because it requires internal changes.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86611
* __->__ #86610
* #86609

bdhirsh committed:
symintify all of derivatives.yaml

[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/NonSymbolicBC.h:
```diff
@@ -10,4 +10,14 @@ namespace native {
+// The below ops don't get a duplicated C++ implementation.
+// They are backward ops, which make them very unlikely to be called directly
+// by external code (at::native::trace_backward).
+// They get their own declaration for BC purposes however.
```
@albanD @wconstab wondering if anyone has thoughts on this. Keeping these extra ops for BC to guarantee no internal failures felt like a good idea. But instead of duplicating all of the code, I just routed them to the symint variants (cuts down on code size a bunch but is marginally slower - shouldn't really matter in practice though, since hopefully nobody is calling these).

bdhirsh commented on aten/src/ATen/templates/NativeFunctions.h:
```diff
@@ -25,6 +25,7 @@
 #include <c10/core/QScheme.h>
 #include <ATen/core/Reduction.h>
 #include <ATen/core/Tensor.h>
+#include <ATen/native/NonSymbolicBC.h>
```
The other change to minimize (guarantee?) no internal breakages is here - fbcode uses `NativeFunctions.h` internally, so they should get the headers for the BC symbols automatically.

albanD reviewed (approved):
Sounds good!
Hopefully it lands cleanly!

bdhirsh commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -3,6 +3,7 @@
 
 #include <ATen/ATen.h>
 #include <ATen/NestedTensorImpl.h>
+#include <ATen/native/NonSymbolicBC.h>
```
I added them to the general `NativeFunctions.h` header (since they're part of `at::native::`) For the OSS builds though, that won't help because we use per-operator headers - so we still need to manually include them here.

That feels a bit weird, but should still accomplish the goal of reducing the risk of internal build failures.

bdhirsh replied:
although - sigh - OSS mobile builds aren't recognizing `ATen/native/NonSymbolicBC.h`. I probably need to add it to a makefile somewhere.

bdhirsh replied:
hmm @albanD - it looks like the build system is complaining because I'm trying to include a header from aten/native (`ATen/native/NonSymbolicBC.h`) inside of a header from plain ATen (`ATen/NativeFunctions.h`).

I think that logically, `NonSymbolicBC.h` and `NativeFunctions.h` should probably live in the same location, since they both refer to native variants of the aten API. That makes me think that `NonSymbolicBC.h` should be moved to live in `aten/src/ATen` instead of `aten/src/ATen/native`.

The whole point of the changes in this PR though were to reduce the risk of internal changes being needed, and that would definitely require an internal change, since that header is used internally now 😛.

So my current plan is just to ignore `NonSymbolicBC.h` entirely in this PR, since it's very unlikely that any of the `at::native::*_backward()` functions that I'm changing are actually directly called anywhere. LMK what you think!

wconstab replied:
sgtm. keep it separate

albanD replied:
Ok!

bdhirsh committed:
Update on "symintify all of derivatives.yaml"

Big-bang PR to symintify **all** .sizes() calls in derivatives.yaml, which will be needed for symbolic tracing.

* with the exception of `split()`, which is tougher to land because it requires internal changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"

Big-bang PR to symintify **all** .sizes() calls in derivatives.yaml, which will be needed for symbolic tracing.

* with the exception of `split()`, which is tougher to land because it requires internal changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "symintify all of derivatives.yaml"

Big-bang PR to symintify **all** .sizes() calls in derivatives.yaml, which will be needed for symbolic tracing.

* with the exception of `split()`, which is tougher to land because it requires internal changes.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge -g
